_posts/2021-08-15-interviews.markdown

```diff
@@ -0,0 +1,155 @@
+
+> In the early 80’s, C programmers were in high demand. Interviewers used questions that specifically tested your experience with C, which meant problems involving lots of pointer manipulation. This ingrained LL questions as a cultural institution in many places, especially places doing lots of low-level work, like Microsoft and Google. From there, it was exported to the wider software world, and lacking the original context, people assumed it was about “testing CS fundamentals” or “quick thinking”.
+
+The fact remains that it's an important part of an interview to make sure that a software engineer can code. So if we're going to be asked to reverse a linked list, what is the real underlying assessment? In 2021, that is probably no longer about pointer manipulation. What should it be about, then? Let's dig in and have some fun.
+
+## Reversing a Linked List
+
+We'll start with the classic algorithm question. My solution is [here](https://gist.github.com/alaiacano/35a63e7631b60641fc0342f31461b80d#file-ll-py). Inserting new nodes at the head of the list is `O(1)` and reversing the list is `O(N)`, and does so by moving pointers around and not making a full copy of the list in memory.
+
+To populate, print, reverse, and then print the list again, we'd do this:
+
+<script src="https://gist.github.com/35a63e7631b60641fc0342f31461b80d.js?file=run_ll.py" type="text/javascript"></script>
+
+Wonderful. In the real world, we would have used [`List.reverse`](https://github.com/python/cpython/blob/e5c8ddb1714fb51ab1defa24352c98e0f01205dc/Objects/listobject.c#L1042-L1056) from the standard library, but this produces the same result. Would we have run the code in a module like this? Probably not!
+
+## Enter the LinkedList Pipeline
+
+Let's rewrite how we use this `LinkedList` class into something a little more _modern_:
+
+<script src="https://gist.github.com/35a63e7631b60641fc0342f31461b80d.js?file=pipeline1.yaml" type="text/javascript"></script>
+
+YAML! Nice. Yaml is fun to joke about, but Vicki Boykis posted [a twitter poll](https://twitter.com/vboykis/status/1407821631997759488) about which languages people use in Data Science/ML and the most common reply was "does yaml count?" If everyone asks, it means yes.
+
+The above yaml does not actually do anything, and this is ostensibly a post about software coding interviews, so let's build something that reverses a linked list given this pipeline definition. Is this overkill for the goal of reversing a 4-element linked list? Sure. But it's a good analog to the modular systems that we are paid to build and use as software engineers (especially Data/ML engineers). It's the kind of thing that we actually do every day, and should be familiar with. In other words, **Components are the new Pointers.**
```

Comment:
I think this point could use a little more elaboration, something like: 

its rare in an actual job in 2021 to write functions like "reverse a linked list" which are provided in standard libraries 99.9% of the time, but what is more common is to assemble various pre-made components together into a larger whole to make something out of them, like this example "pipeline-as-yaml". So lets use this toy example of "reverse a linked list" to demonstrate an interview problem that would have you write a framework for assembling these components into a larger whole (a "pipeline") - since _this_ is likely to more closely resemble the day-to-day work of the job the interview is for.

can also say something like - ok so if this YAML is the interface that LinkedList users will use in our framework, how do we write the code that implements this framework?